pkgs/by-name/im/immich/machine-learning.nix

```
@@ -0,0 +1,85 @@
+{
+  lib,
+  src,
+  immich,
+  python3,
+}:
+python3.pkgs.buildPythonApplication {
+  pname = "immich-machine-learning";
+  inherit (immich) version;
+  src = "${src}/machine-learning";
+  pyproject = true;
+
+  postPatch = ''
+    substituteInPlace pyproject.toml --replace-fail 'fastapi-slim' 'fastapi'
+
+    # Allow immich to use pydantic v2
+    substituteInPlace app/schemas.py --replace-fail 'pydantic' 'pydantic.v1'
+    substituteInPlace app/main.py --replace-fail 'pydantic' 'pydantic.v1'
+    substituteInPlace app/config.py \
+      --replace-fail 'pydantic' 'pydantic.v1'
+  '';
+
+  pythonRelaxDeps = [
+    "setuptools"
+    "pydantic"
+  ];
+  pythonRemoveDeps = [ "opencv-python-headless" ];
+
+  build-system = with python3.pkgs; [
+    poetry-core
+    cython
+  ];
+
+  dependencies =
+    with python3.pkgs;
+    [
+      insightface
+      opencv4
+      pillow
+      fastapi
+      uvicorn
+      aiocache
+      rich
+      ftfy
+      setuptools
+      python-multipart
+      orjson
+      gunicorn
+      huggingface-hub
+      tokenizers
+      pydantic
+    ]
+    ++ uvicorn.optional-dependencies.standard;
+
+  doCheck = false;
+
+  postInstall = ''
+    mkdir -p $out/share
+    cp log_conf.json $out/share
+
+    cp -r ann $out/${python3.sitePackages}/
+
+    makeWrapper ${lib.getExe python3.pkgs.gunicorn} $out/bin/machine-learning \
```

You could also accommodate for an overridable set of `outputs` by using variables such as

```suggestion
    makeWrapper ${lib.getExe python3.pkgs.gunicorn} "''${!outputBin}"/bin/machine-learning \
```

---

IMO this should be added only if/when we need multiple outputs for this package (or e.g: when writing a hook).

---

I'd rather treat phases and ad hoc as normal hooks. This makes it easier to use `.override[Python]Attrs` and simplifies refactoring